text/0000-isa-attribute.md

```diff
@@ -0,0 +1,114 @@
+
+# Motivation
+[motivation]: #motivation
+
+Most programmers are familiar with the idea of a CPU family having more than one instruction set. `x86_64` is backwards compatible with `x86`, and an `x86_64` CPU can run an `x86` program if necessary.
+
+Starting with `ARMv4T`, many ARM CPUs support two separate instruction sets. At the time they were called "ARM code" and "Thumb code", but with the development of `AArch64`, they're now called `a32` and `t32`. Unlike with the `x86` / `x86_64` situation, on ARM you can have a single program that intersperses both `a32` and `t32` code. A particular form of branch instruction allows for the CPU to change between the two modes any time it branches, and so generally code is designated as being either `a32` or `t32` on a per-function basis.
+
+In LLVM, selecting that code should be `a32` or `t32` is done by either disabling (for `a32`) or enabling (for `t32`) the `thumb-mode` target feature. Previously, Rust was able to do this using the `target_feature` attribute because it was able to either add _or subtract_ an LLVM target feature during a function. However, when [RFC 2045](https://github.com/rust-lang/rfcs/blob/master/text/2045-target-feature.md) was accepted, its final form did not allow for the subtraction of target features. Its final form is primarily designed around always opting _in_ to additional features, and it's no longer the correct tool for an "either A or B, but not both" situation like `a32`/`t32` is.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Some platforms support having more than one instruction set used within a single program. Generally, each one will be better for specific parts of a program. Every target has a default instruction set, based on the target triple. If you would like to set a specific function to use an alternate instruction set you use the `#[instruction_set(?)]` attribute, specifying the desired instruction set in parentheses.
+
+Currently this is only of use on ARM family CPUs, which support both the `a32` and `t32` instruction sets. Targets starting with `arm` default to `a32` and targets starting with `thumb` default to `t32`.
+
+```rust
+// this uses the default instruction set for your target
+
+fn add_one(x: i32) -> i32 {
+    x + 1
+}
+
+// This will compile as `a32` code on both `arm` and thumb` targets
+
+#[instruction_set(a32)]
+fn add_five(x: i32) -> i32 {
+    x + 5
+}
+```
+
+To ease the amount of `cfg_attr` required with this attribute, if you specify an instruction set that isn't available on the target used the attribute is simply ignored. For example, if you specify `t32` and then build the code for `x86_64` or `wasm32`, the attribute is ignored.
```

Comment:
If you intend for `instruction_set` to not be solely restricted to ARM then you can't do this. Implicitly ignoring instruction sets that aren't available on a target means that you can't check for typos (`t31` oops) since that might be a valid instruction set name on a different target.

Reply:
Since the most probable names to be used with this are `t32` and `a32`, possibly also `m16`, and `m32` later, that feels unlikely to happen. I think it would be worse to have to put cfg_attr all over the place all the time. I'd be willing to change it if we had to though.